regret/emitted.py

```diff
@@ -1,13 +1,21 @@
 """
 Objects emitted whilst a deprecated object is being used.
 """
+import inspect
+
 import attr
 
 
 def _qualname(obj):
     """
     Return the (non-fully-)qualified name of the given object.
     """
+    if isinstance(obj, str):
```

Comment:
Ech, I see why you needed to add this (to support the module replacement being a string). It makes me a little uneasy to see here but I don't have a better suggestion for the minute...

Reply:
The problem was that it for example I will use `replacement=os.path` it will "resolve" to `ntpath` or `unixpath` depending on the os... so the quick fix was to use `replacement="os.path"` ...

Reply:
Yeah I think it's also reasonable to not want to import the replacement module just to refer to it...

Maybe though instead introducing `Reference(name="os.path")` which just sets `__qualname__` explicitly to what's given is a bit better though? (Name aside, I can't think of a good one at the minute)? The reason is that `_Deprecator` supports some other `qualname` function being used -- e.g. you may want to use `twisted.python.reflect.namedAny` with one. But the type of this has to be well-formed, so asking it to deal with lots of edge cases means other people who want to use it need to deal with them too.

Reply:
I would prefer to keep the `regret` helper call to as many primitive types as possible.

I didn't liked in Twisted that each time I wanted to deprecate something I had to do an extra  import for `incremental`... to mee it feels over-engineering.

I will remov  support for `module` in `_qualname` and kept only extra support for `str`.

My question... why not store `self._replacement` always as `str` ... why do we need to delay resolving to str and call `self._name_of(self._replacement)` ?

Reply:
Well for one as I say you're going to have difficulty passing twisted's existing test suite I think. Twisted shows fully qualified names in warning messages.

Beyond that mostly because if you do already have the object, someone may want to do something with it other than just render it in the string (e.g. if you're writing some library that's processing deprecation warnings that are about to be emitted).

But yeah if it gets in the way there could be things we do instead.